platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiGradleProjectIntegrationTest.groovy

```
@@ -0,0 +1,206 @@
+
+    def "can fetch GradleProject model for empty projects"() {
+        settingsFile << """
+            rootProject.name = 'root'
+
+            include(":lib1")
+            include(":lib1:lib11")
+        """
+
+        when: "fetching without Isolated Projects"
+        def expectedProjectModel = fetchModel(GradleProject)
+
+        then:
+        fixture.assertNoConfigurationCache()
+
+        with(expectedProjectModel) {
+            it.name == "root"
+            it.tasks.size() > 0
+            it.children.size() == 1
+            it.children[0].children.size() == 1
+        }
+
+        when: "fetching with Isolated Projects"
+        executer.withArguments(ENABLE_CLI)
+        def projectModel = fetchModel(GradleProject)
+
+        then:
+        fixture.assertStateStored {
+            modelsCreated(":", 2) // GradleProject, intermediate IsolatedGradleProjectInternal
+            modelsCreated(":lib1", ":lib1:lib11") // intermediate IsolatedGradleProjectInternal
+        }
+
+        checkGradleProject(projectModel, expectedProjectModel)
+
+        when: "fetching again with Isolated Projects"
+        executer.withArguments(ENABLE_CLI)
+        fetchModel(GradleProject)
+
+        then:
+        fixture.assertStateLoaded()
+    }
+
+    def "can fetch GradleProject model without tasks"() {
+        settingsFile << """
+            rootProject.name = 'root'
+
+            include(":lib1")
+        """
+
+        buildFile << """
+            tasks.register("lazy") {
+                println "realizing lazy task"
+            }
+        """
+
+        when: "fetching without Isolated Projects"
+        executer.withArguments("-Dorg.gradle.internal.GradleProjectBuilderOptions=omit_all_tasks")
+        def expectedProjectModel = fetchModel(GradleProject)
+
+        then:
+        fixture.assertNoConfigurationCache()
+        outputDoesNotContain("realizing lazy task")
```

💭 This assertion is a little brittle, as it is easy to accidentally change the build under test so that this ends up testing nothing. I would probably change the `register { }` block to throw an exception instead of logging something.

---

Done